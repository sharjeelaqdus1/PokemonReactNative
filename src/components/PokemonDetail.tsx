import React from 'react';
import {View, StyleSheet, Image, Text} from 'react-native';

import {extractPokemonUrlSegment} from '../helpers/common';
import usePokemonDetail from '../hooks/usePokemonDetail';
import Header from './Header';
import DetailElement from './DetailElement';
import ShouldRender from './ShouldRender';
import DefaultError from './DefaultError';
import {strings} from '../constants/strings';

interface PokemonDetailProps {
  pokemonId: string;
}

const PokemonDetail: React.FC<PokemonDetailProps> = ({pokemonId}) => {
  const {pokemonDetails, error, isLoading} = usePokemonDetail(
    extractPokemonUrlSegment(pokemonId),
  );
  const {name, height, weight, types, sprites} = pokemonDetails || {};

  if (isLoading) {
    return <Text>{strings.loading}</Text>;
  }

  if (error) {
    if ('status' in error) {
      return <DefaultError message={error?.status} />;
    } else {
      return <DefaultError message={strings.failedToLoad} />;
    }
  }

  return (
    <View style={styles.container}>
      <Header showBackButton title={pokemonDetails?.name ?? ''} />
      <View style={styles.detailContainer}>
        <ShouldRender condition={!!sprites?.front_default}>
          <Image
            source={{uri: sprites?.front_default ?? ''}}
            style={styles.image}
          />
        </ShouldRender>
        <ShouldRender condition={!!name}>
          <View style={styles.separator} />
          <DetailElement label={strings.name} value={name} />
        </ShouldRender>
        <ShouldRender condition={!!height}>
          <View style={styles.separator} />
          <DetailElement label={strings.height} value={height + ' cm'} />
        </ShouldRender>
        <ShouldRender condition={!!weight}>
          <View style={styles.separator} />
          <DetailElement label={strings.weight} value={weight + ' kg'} />
        </ShouldRender>
        <ShouldRender condition={!!weight}>
          <View style={styles.separator} />
          <DetailElement label={strings.types} value={types} list={true} />
        </ShouldRender>
      </View>
      <View style={styles.separator2} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,

    backgroundColor: '#fff',
  },
  detailContainer: {
    flex: 1,
    padding: 16,
  },
  image: {
    width: '100%',
    height: 200,
    resizeMode: 'contain',
  },
  name: {
    fontSize: 24,
    fontWeight: 'bold',
    textAlign: 'center',
    marginVertical: 16,
  },
  info: {
    fontSize: 18,
    marginVertical: 8,
  },
  type: {
    fontSize: 16,
  },
  separator: {
    height: 1,
    backgroundColor: '#eee',
    marginVertical: 8,
  },
  separator2: {
    height: 1,
    backgroundColor: 'gray',
    marginVertical: 8,
  },
});

export default PokemonDetail;
