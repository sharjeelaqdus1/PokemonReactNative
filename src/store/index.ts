import {configureStore} from '@reduxjs/toolkit';

import {
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';

import {pokemonApi} from '../services/network';
import {persistedReducer} from './reducer';

export const setupStore = () => {
  const store = configureStore({
    reducer: persistedReducer,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        serializableCheck: false,
        immutableCheck: false,
      }).concat(pokemonApi.middleware),
  });

  return store;
};
export const store = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: false,
      immutableCheck: false,
    }).concat(pokemonApi.middleware),
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
