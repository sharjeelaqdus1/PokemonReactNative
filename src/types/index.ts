export interface Pokemon {
  name: string;
  url: string;
}

export interface PokemonListResponse {
  results: Pokemon[];
}

export interface PokemonDetails {
  id: number;
  name: string;
  height: number;
  weight: number;
  sprites: Sprites;
  types: any[];
  // Add other properties as needed
}

export type Sprites = {
  back_default: string | null;
  back_female: string | null;
  back_shiny: string | null;
  back_shiny_female: string | null;
  front_default: string | null;
  front_female: string | null;
  front_shiny: string | null;
  front_shiny_female: string | null;
};

export type RootStackParamList = {
  Home: undefined;
  Details: {pokemonId: string};
};
