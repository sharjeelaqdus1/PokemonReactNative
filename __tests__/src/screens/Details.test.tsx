import React from 'react';
import {render} from '@testing-library/react-native';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import PokemonDetail from '../../../src/components/PokemonDetail';
import PokemonDetailScreen from '../../../src/screens/Details/Details';
import {RootStackParamList} from '../../../src/types';

jest.mock('../../../src/components/PokemonDetail', () => {
  return jest.fn(() => null);
});

type Props = NativeStackScreenProps<RootStackParamList, 'Details'>;

describe('PokemonDetailScreen', () => {
  const route: Props['route'] = {
    key: 'test-key',
    name: 'Details',
    params: {
      pokemonId: 'pokemon/1/',
    },
  };

  const navigation = {} as Props['navigation'];

  it('renders correctly', () => {
    const {toJSON} = render(
      <PokemonDetailScreen route={route} navigation={navigation} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('passes the correct pokemonId to PokemonDetail component', () => {
    render(<PokemonDetailScreen route={route} navigation={navigation} />);
    expect(PokemonDetail).toHaveBeenCalledWith(
      expect.objectContaining({
        pokemonId: 'pokemon/1/',
      }),
      {},
    );
  });
});
