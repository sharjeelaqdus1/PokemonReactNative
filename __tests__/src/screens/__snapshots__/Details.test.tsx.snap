// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonDetailScreen renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
/>
`;
